### PR TITLE
Document ExtendedMouseButtons pseudo-encoding

### PR DESCRIPTION
**Pull request by CendioHalim**
This pseudo-encoding adds support for additional mouse buttons, most notably the forward and backward mouse buttons.
* Open PR: https://github.com/TigerVNC/tigervnc/pull/1826

### CODE REVIEW

**Comment by CendioHalim** [comment]
The pseudo-encoding has been registered with [IANA](https://www.iana.org/assignments/rfb/rfb.xhtml#rfb-3) and has the assignment:

 **-316 	ExtendedMouseButtons**

**Review by samhed — APPROVED** [review]

File: rfbproto.rst [hunk 1]
```diff
@@ -1674,6 +1674,9 @@ The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
 negotiation of an alternative interpretation for the *x-position*
 and *y-position* fields, as relative deltas.
 
+The `ExtendedMouseButtons Pseudo-encoding`_ extends the mouse button
+functionality by adding an additional byte to the *button-mask*.
+
```
[the remark this answers is not in the document]
Comment (CendioHalim) [reply rfbproto.rst:1679]: This is already written, the last paragraph at the bottom:
```
Currently, bit 1 in the *extended-button-mask* corresponds to the
forward mouse button, and bit 2 corresponds to the back mouse button.
The remaining 6 bits are reserved.
```

Do you think there should be a similar text added here as well, part of the PointerEvent chapter?

[the remark this answers is not in the document]
Comment (berrange) [reply rfbproto.rst:1679]: Oh, that's easy to miss at a glance. If we present the information in each chapter as a table rather than prose, I think it'd be more visible. ie in this chapter add

```
=== ==========================
Bit   Button
=== ==========================
0   Left
1   Middle
2   Right 
3   Wheel Up
4   Wheel Down
5   Wheel Left
6   Wheel Right
7   ExtendedPointerEvent Marker
=== =========================
```

And in the latter (new) chapter  add

```
=== ==========================
Bit   Button
=== ==========================
0   Forward
1   Back
2   Reserved 
3   Reserved
4   Reserved
5   Reserved
6   Reserved
7   Reserved
=== =========================
```

[the remark this answers is not in the document]
Comment (CendioHalim) [reply rfbproto.rst:1679]: That's a good point. It's much easier and quicker to read when presented as a table

File: rfbproto.rst [hunk 2]
```diff
@@ -4088,6 +4092,73 @@ include the extra bits that are used for alpha. Also note that the data
...
+========= =============================================================
+0         Forward
+1         Back
+2         Reserved
+3         Reserved
+4         Reserved
+5         Reserved
+6         Reserved
+7         Reserved
+========= =============================================================
```
Comment (berrange) [inline rfbproto.rst:4160]: This is good, but this commit is missing the companion table under "PointerEvent"

Comment (CendioHalim) [reply rfbproto.rst:4179]: Added a table to "PointerEvent" as well, but I'm not sure the placement is the best. What do you think?

Comment (berrange) [reply rfbproto.rst:4179]: It is ok by me.

File: rfbproto.rst [hunk 3]
```diff
@@ -4088,6 +4108,74 @@ include the extra bits that are used for alpha. Also note that the data
 used for the cursor shares state with other rects. E.g. the zlib stream
 for a ZRLE encoding is the same as for data rects.
 
+ExtendedMouseButtons Pseudo-encoding
+------------------------------------
```
Comment (CendioOssman) [inline rfbproto.rst:4112]: Most of this chapter describes the contents of the `PointerEvent` message. I think it's better if those portions are in that message's chapter, as that's where people will look for how it should be handled.

I would suggest keeping just the first paragraph. Possibly also just referring to it as an "extended PointerEvent message" and defining what that means in the other chapter.

File: rfbproto.rst [hunk 4]
```diff
@@ -4088,6 +4108,74 @@ include the extra bits that are used for alpha. Also note that the data
...
+2               ``U16``                         *y-position*
+1               ``U8``                          *extended-button-mask*
+=============== ==================== ========== =======================
+
+While a normal `PointerEvent`_ uses 7 of the 8 available bits, the
+*ExtendedMouseButtons* pseudo-encoding provides support for 8 additional
+bits, for a total of 15 buttons. To allow for 8 additional bits of data
+being transmitted using the same `PointerEvent`_  message, a marker bit
+is used to indicate that there is an extra byte of data in the
+`PointerEvent`_  message. The marker bit is the 7th bit of
```
Comment (CendioOssman) [inline rfbproto.rst:4137]: The 7th bit is bit 6, which is probably not what you mean. :)

Given the system here, I would suggest calling it the highest bit. Least risk of confusion.

File: rfbproto.rst [hunk 5]
```diff
@@ -4088,6 +4108,74 @@ include the extra bits that are used for alpha. Also note that the data
...
+mask.
+
+When decoding the `PointerEvent`_ message with this pseudo-encoding
+enabled, it can initially be decoded as a normal `PointerEvent`_
+message. After decoding the message, the 7th bit in the *button-mask* is
+used to determine whether an additional byte should be read or not. If
+the marker bit is set to one, another byte is read, which contains the
+*extended-button-mask*. The last step in decoding is to set the 7th bit
+of the *button-mask* is to zero.
+
```
Comment (CendioOssman) [inline rfbproto.rst:4157]: You may not need to be this verbose. It's sufficient to clearly state how the messages are expected to look like. Then it's up to the implementation how to send/receive them.

File: rfbproto.rst [hunk 6]
```diff
@@ -4088,6 +4108,74 @@ include the extra bits that are used for alpha. Also note that the data
...
+If the marker bit in  the *button-mask* is set to zero, the message is a
+normal `PointerEvent`_ message and, no further action is needed.
+
+Currently, bit 0 in the *extended-button-mask* corresponds to the
+forward mouse button, and bit 1 corresponds to the back mouse button.
+The remaining 6 bits are reserved and must be set to 0 by the client and
+ignored by the server.
```
Comment (CendioOssman) [inline rfbproto.rst:4164]: As discussed elsewhere, this may need a warning/clarification on what we consider "forward" and "backward".

I suggested a dive in to the USB specification to increase the clarity of this mess. Has that been done?

File: rfbproto.rst [hunk 7]
```diff
@@ -4088,6 +4108,74 @@ include the extra bits that are used for alpha. Also note that the data
...
+Bit       Description
+========= =============================================================
+0         Forward
+1         Back
+2         Reserved
+3         Reserved
+4         Reserved
+5         Reserved
+6         Reserved
+7         Reserved
```
Comment (CendioOssman) [inline rfbproto.rst:4176]: It's best to specify what reserved means for the client and server. I would suggest clients are required to send 0, and servers are required to ignore them.

File: rfbproto.rst [hunk 8]
```diff
@@ -1674,6 +1674,25 @@ The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
 negotiation of an alternative interpretation for the *x-position*
 and *y-position* fields, as relative deltas.
 
+The `ExtendedMouseButtons Pseudo-encoding`_ extends the mouse button
+functionality by adding an additional byte for button states. When
+this pseudo-encoding is enabled, bit 7 in the *button-mask* can be set
+to 1. Clients that don't use this pseudo-encoding must set the marker
+bit to 0.
```
Comment (CendioOssman) [inline rfbproto.rst:1681]: I think we need to consider old/current clients more here.

What happens if a client sends the high bit set, but no ExtendedMouseButtons pseudo-encoding?

And clients supporting ExtendedMouseButtons likely need to be forced to send the high bit 0 until they know the server supports it?

**Review by CendioOssman — CHANGES_REQUESTED** [review]
Looks promising! Some clarifications and we should be good.

**Comment by RokerHRO** [comment]
I'd favour a more flexible and extensible approach, that would also work for so-called "gaming" mice with lots of additional buttons (that might be mapped to keybord shortcts at server side) and for other USB input devices such as styluses or digitizer tablets that are able to report ''strength'' of pressure etc.
 
Maybe a look into the USB HID specification would help, so we don't have to re-invent the wheel, because USB HID already covers other extended "event input" devices.

**Comment by any1** [comment]
> I'd favour a more flexible and extensible approach

Yes, let's also do that, but this is fine as is.

> Maybe a look into the USB HID specification

That's actually a pretty good idea. USB is really just a transport for HID. I know of at least one non-USB use of HID: Miracast passes raw HID events (and descriptors) back to the presenter. This is mainly useful with touch panels.

I look forward to seeing proposals for VNC-HID. :)

**Comment by CendioOssman** [comment]
> I'd favour a more flexible and extensible approach, that would also work for so-called "gaming" mice with lots of additional buttons (that might be mapped to keybord shortcts at server side)

That is something we thought of as well. Turns out, these mice (at least not the ones we have access to) do not send mouse events. They send keyboard events for all those extra buttons. :/

My guess would be because Windows doesn't have a convenient interface for anything more than the back and forward button, so why send hardware events that most applications cannot easily use.

> and for other USB input devices such as styluses or digitizer tablets that are able to report ''strength'' of pressure etc.

That would be nice, but is a much bigger project. Might also require bidirectional signalling to reconfigure things. 

Let's not have those grander ambitions block fixing a common use case like browser interaction.

File: rfbproto.rst (hunk 8 above)
Comment (CendioHalim) [reply rfbproto.rst:1681]: > What happens if a client sends the high bit set, but no ExtendedMouseButtons pseudo-encoding

I tested this, using a client that sent a button mask with the high bit set, with other servers (https://github.com/TigerVNC/tigervnc/pull/1711#issuecomment-2145069550).

I should probably test against more implementations to see if there is anything that breaks, but from what I've tested, setting the high bit only affected the RealVNC server, which corresponded with a back button press.

> And clients supporting ExtendedMouseButtons likely need to be forced to send the high bit 0 until they know the server supports it?

Yes, I will add that.

Comment (CendioOssman) [reply rfbproto.rst:1681]: > I tested this, using a client that sent a button mask with the high bit set, with other servers ([TigerVNC/tigervnc#1711 (comment)](https://github.com/TigerVNC/tigervnc/pull/1711#issuecomment-2145069550)).

What I meant was, what happens when you do this to a server that *supports* this extension? It might expect a second byte that never comes.

File: rfbproto.rst (hunk 7 above)
Comment (CendioHalim) [reply rfbproto.rst]: Good point!

File: rfbproto.rst (hunk 8 above)
Comment (CendioHalim) [reply rfbproto.rst]: I've updated the spec so that servers that support this extension ignore the marker bit unless the client also supports it. Clients are also required to send the marker bit as 0 if they support the extension, until the server has indicated that it also supports the extension.

File: rfbproto.rst (hunk 6 above)
Comment (CendioHalim) [reply rfbproto.rst]: As discussed, I agree it's a good idea. I haven't looked at the USB HID specification yet.

**Comment by RokerHRO** [comment]
Am I right: this protocol extension enables only 2 more mouse buttons (with pre-defined meanings)?

So if a future client wants to support a super fancy mouse with another cool button there is _no way_ for a client to use the "reserved" bits for it nor to negotiate with the server the presence and the meaning of the remaining bits. Another crude "pseudo encoding" hack has to be defined in that case.

**Comment by CendioHalim** [comment]
> Am I right: this protocol extension enables only 2 more mouse buttons (with pre-defined meanings)?

Correct. There are two motivations for this:

1. Simplicity - This is very easy to implement on both the client and server side.
2. QoL - For me personally, this addition is a huge QoL improvement for VNC. While a more general solution would be very cool to have, it would require significant effort to implement. The work needed to implement vs. actual usability, with the current solution, is very high in my opinion.

> So if a future client wants to support a super fancy mouse with another cool button there is _no way_ for a client to use the "reserved" bits for it nor to negotiate with the server the presence and the meaning of the remaining bits. Another crude "pseudo encoding" hack has to be defined in that case.

The point of leaving the remaining bits reserved is to leave room for future additions to this pseudo-encoding. Future additions could be either more buttons, or other hacks for a more general approach (or another pseudo-encoding entirely). 

Having the unused bits reserved as of now should make it easier to preserve backwards compatibility.

File: rfbproto.rst [hunk 9]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
 1               ``U8``                          *button-mask*
 2               ``U16``                         *x-position*
 2               ``U16``                         *y-position*
+1               ``U8``                          *extended-buttons-mask*
```
Comment (CendioOssman) [inline rfbproto.rst:1671]: I don't think we have any precedence of having optional fields like this in a description. I'm worried someone might read this too casually and think this field is always present.

What are your thoughts on keeping the original table, and then having an almost duplicate table showing the extended mode?

File: rfbproto.rst [hunk 10]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+========= =============================================================
+0         Left
+1         Middle
+2         Right
+3         Wheel-up
+4         Wheel-down
+5         Wheel-left
+6         Wheel-right
+7         ExtendedMouseButtons Marker
+========= =============================================================
```
Comment (CendioOssman) [inline rfbproto.rst:1689]: This table duplicates what's written earlier. It would probably be best to merge things in some way.

File: rfbproto.rst [hunk 11]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+Bit       Description
+========= =============================================================
+0         Left
+1         Middle
+2         Right
+3         Wheel-up
+4         Wheel-down
+5         Wheel-left
+6         Wheel-right
+7         ExtendedMouseButtons Marker
```
Comment (CendioOssman) [inline rfbproto.rst:1688]: This is only well-defined with this new extension. It is probably best to mention that it has no good historical definition. Or did we find any client or server that used it? I would assume it would be the "back" button if used.

File: rfbproto.rst [hunk 12]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+========= =============================================================
 
 The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
 negotiation of an alternative interpretation for the *x-position*
 and *y-position* fields, as relative deltas.
 
+The `ExtendedMouseButtons Pseudo-encoding`_ extends the mouse button
+functionality by adding an additional byte for button states. When
+this pseudo-encoding is enabled, the highest bit in the *button-mask*
+can be set to 1.
```
Comment (CendioOssman) [inline rfbproto.rst:1698]: **must** be set to 1 if this extra byte is used.

File: rfbproto.rst [hunk 13]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+7         ExtendedMouseButtons Marker
+========= =============================================================
 
 The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
 negotiation of an alternative interpretation for the *x-position*
 and *y-position* fields, as relative deltas.
 
+The `ExtendedMouseButtons Pseudo-encoding`_ extends the mouse button
+functionality by adding an additional byte for button states. When
+this pseudo-encoding is enabled, the highest bit in the *button-mask*
```
Comment (CendioOssman) [inline rfbproto.rst:1697]: I'm concerned that this phrasing could be interpreted as it being sufficient to do `SetEncodings` with `ExtendedMouseButtons`.

Perhaps use the phrase "this extension is enabled" and readers will need to look at the other chapter for details?

File: rfbproto.rst [hunk 14]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+*ExtendedMouseButtons* pseudo-encoding provides support for 8 additional
+bits, for a total of 15 buttons. To allow for 8 additional bits of data
+being transmitted using the same `PointerEvent`_  message, a marker bit
+is used to indicate that there is an extra byte of data in the
+`PointerEvent`_  message. The marker bit is the highest of the
+*button-mask*, where a value of one means that there is an additional
+byte of data to read, and a value of zero means that it is a normal
+`PointerEvent`_ message.
+
+Clients that don't support this pseudo-encoding must set the highest bit
```
Comment (CendioOssman) [inline rfbproto.rst:1710]: We can't retroactively change client behaviour.

File: rfbproto.rst [hunk 15]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+in *button-mask* to 0, and servers that support this extension must
+ignore the marker bit if the client does not support it. If the server
```
Comment (CendioOssman) [inline rfbproto.rst:1712]: "it" here is ambiguous. Should probably explicitly state "this extension".

Comment (CendioOssman) [inline rfbproto.rst:1712]: Is this perhaps a bit harsh? If the client has not signalled support, then shouldn't legacy behaviour persist?

File: rfbproto.rst [hunk 16]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+does not support this pseudo-encoding, clients must set the marker bit
+to 0.
```
Comment (CendioOssman) [inline rfbproto.rst:1714]: I think it's better to state "Until the client knows the server supports this extension...". It is not just about legacy servers, but also a race until the handshake is complete.

File: rfbproto.rst [hunk 17]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+
+An extended `PointerEvent`_  message is encoded by first setting the
+highest bit of the *button-mask* to one. The *button-mask* will contain
+the button mask for the 7 buttons defined in a normal PointerEvent`_
+message, as well as the marker bit. The rest of the message is sent as a
+normal `PointerEvent`_ message. After sending the *y-position*, an
+additional byte, the *extended-buttons-mask*, is sent containing a one
+byte button mask.
```
Comment (CendioOssman) [inline rfbproto.rst:1722]: Part of this can probably be removed, as it should be clear from the table. In order to keep the section from getting too long.

File: rfbproto.rst [hunk 18]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+
+When decoding the `PointerEvent`_ message with this pseudo-encoding
+enabled, it can initially be decoded as a normal `PointerEvent`_
+message. After decoding the message, the highest bit in the
+*button-mask* is used to determine whether an additional byte should be
+read or not. If the marker bit is set to one, another byte is read,
+which contains the *extended-buttons-mask*.
+
+If the marker bit in  the *button-mask* is set to zero, the message is a
+normal `PointerEvent`_ message and, no further action is needed.
```
Comment (CendioOssman) [inline rfbproto.rst:1732]: It is probably best to explicitly state that any high bits are assumed to be zero when this happens. Otherwise, it becomes unclear what it means to send an extended event with high mouse buttons pressed, followed by a normal event.

File: rfbproto.rst [hunk 19]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+normal `PointerEvent`_ message. After sending the *y-position*, an
+additional byte, the *extended-buttons-mask*, is sent containing a one
+byte button mask.
+
+When decoding the `PointerEvent`_ message with this pseudo-encoding
+enabled, it can initially be decoded as a normal `PointerEvent`_
+message. After decoding the message, the highest bit in the
+*button-mask* is used to determine whether an additional byte should be
+read or not. If the marker bit is set to one, another byte is read,
+which contains the *extended-buttons-mask*.
```
Comment (CendioOssman) [inline rfbproto.rst:1729]: I wonder if this is a bit too step-by-step and hence gets a bit wordy and difficult to read. We should describe the wire format, not how the code should be written.

Comment (CendioOssman) [inline rfbproto.rst:1729]: Somewhere in here, it needs to be explicit that this depends on whether the client has advertised support for the extension or not. Otherwise, a high bit does not mean that there is more data.

File: rfbproto.rst [hunk 20]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+
+If the marker bit in  the *button-mask* is set to zero, the message is a
+normal `PointerEvent`_ message and, no further action is needed.
+
+Currently, bit 0 in the *extended-buttons-mask* corresponds to the
+backward mouse button, and bit 1 corresponds to the forward mouse
```
Comment (CendioOssman) [inline rfbproto.rst:1735]: This is redundant with the table

File: rfbproto.rst [hunk 21]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+button. The remaining 6 bits are reserved and must be set to 0 by the
+client and ignored by the server.
```
Comment (CendioOssman) [inline rfbproto.rst:1737]: This is good and should still be included. It allows defining those bits at a later date.

File: rfbproto.rst [hunk 22]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+
+Note that on X11 platforms, the back and forward mouse buttons are
+represented by the labels BTN_LABEL_PROP_BTN_SIDE and
+BTN_LABEL_PROP_BTN_EXTRA respectively.
```
Comment (CendioOssman) [inline rfbproto.rst:1741]: Those are rather internal constants and are probably not too useful to most people.

I think it is sufficient to write something generic that there has historically been some confusion on which buttons are Back/Forward. And perhaps instead define them by behaviour. I.e. what they do in applications.

And since it is an implementation hint, it might be best as a note.

File: rfbproto.rst [hunk 23]
```diff
@@ -1668,12 +1668,93 @@ No. of bytes    Type                 [Value]    Description
...
+Bit       Description
+========= =============================================================
+0         Back
+1         Forward
+2         0
+3         0
+4         0
+5         0
+6         0
+7         0
```
Comment (CendioOssman) [inline rfbproto.rst:1755]: `0` suggests that it will always be the only value. "Reserved" is more clear what is going on here.

File: rfbproto.rst [hunk 24]
```diff
@@ -4088,6 +4170,21 @@ include the extra bits that are used for alpha. Also note that the data
...
+
+A client which requests the *ExtendedMouseButtons* pseudo-encoding is
+declaring that it is capable of handling additional mouse buttons.
+Clients can notify a server that it supports *ExtendedMouseButtons* by
+sending a `SetEncodings`_ message with the *ExtendedMouseButtons*
+pseudo-encoding.
+
+Servers that support this pseudo-encoding will respond to a client's
+request by sending an empty pseudo-rectangle with the matching
+pseudo-encoding. When a client has received the matching response from
```
Comment (CendioOssman) [inline rfbproto.rst:4184]: *When* will the server send this? It's best to be explicit.

File: rfbproto.rst (hunk 12 above)
Comment (CendioHalim) [reply rfbproto.rst:1698]: Hmm, my reasoning was that you don't **have** to send an extended `PointerEvent` message unless you have to (to save bandwidth or whatever).  The way it is worded now is a bit ambiguous.

Changing it to **must** would make things easier, and the only drawback is that an additional byte is sent per `PointerEvent` message. 

What do you think is best? Forcing all  `PointerEvent`  messages to be extended with this pseudo-encoding, or only forcing it if you want to send the extra byte?

File: rfbproto.rst (hunk 11 above)
Comment (CendioHalim) [reply rfbproto.rst:1688]: I tested the following servers:
  * TurboVNC
  * x0vncserver
  * RealVNC
  * libvncserver
    * x11vnc
 
Out of these, only x0vncserver responded to the 7th bit being set, and I was able to navigate "back" on Firefox/Chrome.
 
 I tested the following clients with an Xvnc server that looks at the 7th bit:
   * TurboVNC
   * TightVNC
   * RealVNC
   * Remmina
   * noVNC

Out of these, only RealVNC responded to clicking on the back button.

Comment (CendioHalim) [reply rfbproto.rst:1688]: I guess this means that the current wording:
```While a normal `PointerEvent`_ uses 7 of the 8 available bits```
is incorrect. The back button is nothing that has ever been a part of the specification, my guess is that the edge cases I've seen are purely coincidental.

File: rfbproto.rst (hunk 8 above)
Comment (CendioHalim) [reply rfbproto.rst]: I thought about this one more time, and I think that the behaviour should be the following:

| Server pseudo-encoding   | Client pseudo-encoding | Behaviour |
| ------------------------------------- | ----------------------------------- | --------------- |
| No                                         | No                                        | Client sends 1 byte, highest bit **can** be set to indicate back button. <br> Server reads 1 byte, **don't** ignore highest bit, **can** interpret highest bit as back button.|
| Yes                                         | No                                        | Client sends 1 byte, highest bit **can** be set to indicate back button. <br> Server reads 1 byte, **don't** ignore highest bit, **can** interpret highest bit as back button.|
| No                                         | Yes                                        | Client sends 1 byte, highest bit **can** be set to indicate back button. <br> Server reads 1 byte, **don't** ignore highest bit, **can** interpret highest bit as back button.|
| Yes                                         | Yes                                        | Client sends 1 byte, highest bit **can** be set to indicate extended `PointerEvent`.  If highest bit is set, another byte **must** be sent.<br> Server reads 1 byte and **must** read highest bit. If highest bit is set, server **must** read another byte.|

This way, we should preserve the old behaviour except for in the case where both the client and server have indicated the pseudo-encoding support, and the client sets the highest bit.

Servers should **only** expect another byte if the client has indicated pseudo-encoding support, and the highest bit is set.

File: rfbproto.rst [hunk 25]
```diff
@@ -1674,6 +1674,67 @@ The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
...
+No. of bytes    Type                 [Value]    Description
+=============== ==================== ========== =======================
+1               ``U8``               5          *message-type*
+1               ``U8``                          *button-mask*
+2               ``U16``                         *x-position*
+2               ``U16``                         *y-position*
+1               ``U8``                          *extended-buttons-mask*
+=============== ==================== ========== =======================
+
+The *ExtendedMouseButtons* pseudo-encoding provides support for 8
```
Comment (CendioOssman) [inline rfbproto.rst:1693]: Technically, 7 additional?

File: rfbproto.rst [hunk 26]
```diff
@@ -1674,6 +1674,67 @@ The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
...
+
+Bit 0 in *extended-buttons-mask* represents the back button on a mouse,
+and bit 1 represents the forward button (see [NOTE2]_ for additional
+context). Bits 2-7 are reserved and must be set to 0 by the client and
+ignored by the server.
+
+========= =============================================================
+Bit       Description
+========= =============================================================
+0         Back
```
Comment (CendioOssman) [inline rfbproto.rst:1722]: I guess we leave the button 8 for the standard message undefined for now?

File: rfbproto.rst [hunk 27]
```diff
@@ -1674,6 +1674,67 @@ The `QEMU Pointer Motion Change Pseudo-encoding`_ allows for the
...
+
+The *ExtendedMouseButtons* pseudo-encoding provides support for 8
+additional button states, giving support for a total of 15 buttons.
+To allow for 8 additional bits of data being transmitted using the same
+`PointerEvent`_ message, a marker bit is used to indicate that there is
+an extra byte of data in the `PointerEvent`_ message. The marker bit is
+the highest bit of the *button-mask*. Clients that support this
+pseudo-encoding must always set the marker bit to zero until the server
+indicates that it supports extended `PointerEvent`_ messages.
+
```
Comment (CendioOssman) [inline rfbproto.rst:1701]: It's still not explicitly stated what happens if you mix standard and extended messages.

**Review by CendioOssman — CHANGES_REQUESTED** [review]

File: rfbproto.rst (hunk 27 above)
Comment (berrange) [reply rfbproto.rst:1701]: From the earlier text 

```
 When this pseudo-encoding is enabled, the highest bit in 
the *button-mask* can be set to one, which marks the `PointerEvent`_ 
message as an extended `PointerEvent`_ message"
```

I interpreted it was permitted to mix both standard & extended messages at the clients discretion, since the word `can` was used rather than `must`. Lets make the intention explicit though.

File: rfbproto.rst (hunk 26 above)
Comment (berrange) [reply rfbproto.rst:1722]: Hmm, interesting question.   What happens if we have a future situation where some VNC server/client that's unaware of this extension, defines a mapping for button 8 in the standard message. If it later wants to use the extended message, it can't use button 8 in the standard message anymore, as that bit is used to indicate that extended message. We would nee to re-map it into the extended message I guess ?

Comment (CendioOssman) [reply rfbproto.rst:1722]: I think that ship has sailed, given that RealVNC and x0vncserver already have some behaviour for that bit. So it's more that we haven't explicitly documented it.

Comment (berrange) [reply rfbproto.rst:1722]: Hmm, awkward. I can't help thinking that rather than having the encoding alter the semantics of the PointerEvent message, it would be simpler if we requested assignment of a "ExtendedPointerEvent" message, and then just defined its button-mask as a `u16`. That way we wouldn't have to assign magic semantics to bit 8, which clash with pre-existing usage of bit 8.

Comment (CendioOssman) [reply rfbproto.rst:1722]: Indeed. But the namespace for messages is a lot smaller than for encodings. So they idea here was to avoid an allocation if possible.

Comment (CendioOssman) [reply rfbproto.rst:1722]: And it's not like that button is somehow lost. It just gets shifted up to the lowest bit in the extended byte.

File: rfbproto.rst (hunk 27 above)
Comment (CendioHalim) [reply rfbproto.rst:1701]: > I interpreted it was permitted to mix both standard & extended messages at the clients discretion, since the word `can` was used rather than `must`. Lets make the intention explicit though.

That was my reasoning, but yeah this could be described more explicitly.

File: rfbproto.rst (hunk 26 above)
Comment (CendioHalim) [reply rfbproto.rst:1722]: > Hmm, interesting question. What happens if we have a future situation where some VNC server/client that's unaware of this extension, defines a mapping for button 8 in the standard message. If it later wants to use the extended message, it can't use button 8 in the standard message anymore, as that bit is used to indicate that extended message. We would nee to re-map it into the extended message I guess ?

> And it's not like that button is somehow lost. It just gets shifted up to the lowest bit in the extended byte.

Indeed, button 8 would be assigned to the first bit in the extended mask.

The only thing that we break with this pseudo-encoding is that clients that want to enable this feature are **not** allowed to send bit 8 when sending normal `PointerEvent` messages. The damage is quite minimal in my opinion, as there are barely any servers/clients that use bit 8 anyway and the bit's meaning is not defined in the protocol.

Comment (CendioHalim) [reply rfbproto.rst:1722]: > I guess we leave the button 8 for the standard message undefined for now?

I don't feel confident that we can/should define button 8, undefined is probably best.

File: rfbproto.rst (hunk 27 above)
Comment (CendioHalim) [reply rfbproto.rst:1701]: Or perhaps using the word `may`, which is used many times in the specification already? It also matches [RFC2119](https://datatracker.ietf.org/doc/html/rfc2119) more closely.

File: rfbproto.rst (hunk 25 above)
Comment (CendioHalim) [reply rfbproto.rst:1693]: Hmm, depends. If we specify bit 8 in the normal message as undefined, then we do technically get 8 more bits to play with.

File: rfbproto.rst (hunk 11 above)
Comment (CendioHalim) [reply rfbproto.rst]: > Out of these, only RealVNC responded to clicking on the back button.

My mistake, what I saw was that RealVNC server `vncserver-x11`, (not `vncserver-virtual`) responded on clicking the back button.

**Review by CendioOssman — CHANGES_REQUESTED** [review]
Looks pretty good. Some clarifications with the last changes could be good, then we can merge.

**Review by CendioOssman — APPROVED** [review]
